### PR TITLE
[SPARK-13601] call failure callbacks before writer.close()

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

In order to tell OutputStream that the task has failed or not, we should call the failure callbacks BEFORE calling writer.close().
## How was this patch tested?

Added new unit tests.

### CODE REVIEW

**Comment by davies** [comment]
cc @rxin

**Comment by rxin** [comment]
Thanks for doing this. I took a look and I think we also need to update a few of the places in WriterContainer.scala, because the SQL file-based data sources don't actually go through this pair rdd path.

**Comment by davies** [comment]
@rxin  Added.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -1267,6 +1267,45 @@ private[spark] object Utils extends Logging {
...
+   * exception.
+   *
+   * This is primarily an issue with `finally { out.close() }` blocks, where
+   * close needs to be called to clean up `out`, but if an exception happened
+   * in `out.write`, it's likely `out` may be corrupted and `out.close` will
+   * fail as well. This would then suppress the original/likely more meaningful
+   * exception from the original `out.write` call.
+   */
+  def tryWithSafeFinallyAndFailureCallbacks[T](block: => T)(finallyBlock: => Unit): T = {
+    // It would be nice to find a method on Try that did this
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: remove this comment

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 2]
```diff
@@ -263,16 +261,17 @@ private[sql] class DefaultWriterContainer(
     } catch {
       case cause: Throwable =>
         logError("Aborting task.", cause)
+        // call failure callbacks first, so we could have a chance to cleanup the writer.
+        TaskContext.get().asInstanceOf[TaskContextImpl].markTaskFailed(cause)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: maybe more clear if we move this into abortTask?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: actually never mind - more clear here

**Comment by rxin** [comment]
LGTM pending tests.

Can you just remove that comment when you merge it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala (hunk 2 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: We can't access `cause` in abortTask.

**Comment by davies** [comment]
Merging this into master (comments removed).

**Pull request closed without merging**